### PR TITLE
Fix: error handling in function runner

### PR DESCRIPTION
**Pull request by nickpetrovic**
- Marks the task status as error when it fails
- Fixes or prevents container from running until task expires e.g. "container still running"
- Fixes issue where CLI never exits - still mentions "Waiting..." even though the error has been printed
- Fixes scoping with some vars

Resolve BE-1904

### CODE REVIEW

File: sdk/src/beta9/runner/function.py [hunk 1]
```diff
@@ -131,87 +125,120 @@ def main(channel: Channel):
     gateway_stub: GatewayServiceStub = GatewayServiceStub(channel)
     task_id = config.task_id
 
-    with StdoutJsonInterceptor(task_id=task_id), ThreadPoolExecutorOverride() as thread_pool:
+    if not task_id:
+        raise RunnerException("Invalid runner environment")
+
+    start_time = time.time()
```
Comment (jsun-m) [inline sdk/src/beta9/runner/function.py:131]: Why did  we need to move the start_time out?

File: sdk/src/beta9/runner/function.py [hunk 2]
```diff
@@ -131,87 +125,120 @@ def main(channel: Channel):
...
-                task_id=task_id,
-                task_duration=task_duration,
-                task_status=task_status,
-                container_id=container_id,
-                container_hostname=container_hostname,
-                keep_warm_seconds=0,
-            ),
-            override_callback_url=kwargs.get("callback_url"),
-        )
+def start_task(gateway_stub, task_id, container_id):
```
Comment (jsun-m) [inline sdk/src/beta9/runner/function.py:176]: did this functioanlly change anything?

**Review by jsun-m — CHANGES_REQUESTED** [review]
This looks much cleaner. However what was the issue that this is solving in terms of the hanging function? How does moving the threadpoolexecutor further into the scope of the log interceptor help aid it in shutting down a container?

File: sdk/src/beta9/runner/function.py (hunk 2 above)
Comment (nickpetrovic) [reply sdk/src/beta9/runner/function.py:175]: No, but it works with the theme of having the functions described as invoke, handle error, and complete task. It looks cleaner is all. Functionally, this does not worsen the code, it just makes it easier to read.

Comment (jsun-m) [reply sdk/src/beta9/runner/function.py:175]: makes sense

File: sdk/src/beta9/runner/function.py [hunk 3]
```diff
@@ -131,87 +125,120 @@ def main(channel: Channel):
...
+            try:
+                result = invoke_function(function_stub, context, task_id)
+            except BaseException as e:
+                handle_task_failure(e, gateway_stub, task_id, container_id, container_hostname)
+                raise
+            finally:
+                monitor_future.cancel()
+
+            # End the task and send callback
+            complete_task(
```
Comment (dleviminzi) [inline sdk/src/beta9/runner/function.py:165]: Is there no need to handle exceptions on these calls? If we fail to complete the task should we re-try?

Comment (nickpetrovic) [reply sdk/src/beta9/runner/function.py:173]: No, its what I thought too but found out that `end_task_and_send_callback` (which is called by complete_task) has this decorator on it.

```python
@retry_grpc_call(exception_to_check=BaseException, tries=4, delay=5, backoff=2)

```

Comment (nickpetrovic) [reply sdk/src/beta9/runner/function.py:173]: I mean we might still need it, but I don't knwo what we'd do with the error since we can complete the task.

Comment (dleviminzi) [reply sdk/src/beta9/runner/function.py:173]: Ahh I see

File: sdk/src/beta9/runner/function.py (hunk 1 above)
Comment (nickpetrovic) [reply sdk/src/beta9/runner/function.py]: Just moved a lot around. Its moved back to where it was.

**Review by jsun-m — APPROVED** [review]

**Pull request merged**